File: hyprbar/modules/calendar.py
from gi.repository import Gtk, GLib
from hyprbar.widgets.module_container import ModuleContainer
from datetime import datetime


class CalendarPopoverContent(Gtk.Box):
    def __init__(self):
        super().__init__()
        self.set_orientation(Gtk.Orientation.VERTICAL)
        calendar = Gtk.Calendar()
        self.time_label = Gtk.Label(label="13 : 56")
        self.time_label.get_style_context().add_class("title-1")
        self.time_label.get_style_context().add_class("big-date-time")
        self.date_label = Gtk.Label(label="Wednesday, 24 August")
        self.date_label.get_style_context().add_class("title-4")
        self.date_label.get_style_context().add_class("full-date")
        self.append(self.time_label)
        self.append(self.date_label)
        self.append(calendar)


class Calendar(ModuleContainer):
    def __init__(self, config, window):
        super().__init__(config, window, is_button=True)

        self.popover_content = CalendarPopoverContent()

        self.popover = Gtk.Popover()
        self.popover.set_child(self.popover_content)
        self.popover.set_position(Gtk.PositionType.BOTTOM)
        self.popover.set_has_arrow(False)
        self.popover.set_offset(0, 8)
        self.popover.set_parent(self.container)
        self.container.connect("clicked", self.on_click)

        self.label = Gtk.Label()
        self.update_date_time()
        self.set_child(self.label)

    def on_click(self, window):
        self.popover.popup()

    def update_date_time(self):
        now = datetime.now()
        self.label.set_markup(f"<b>{now.strftime('%d %b  %I:%M %p')}</b>")
        self.popover_content.time_label.set_label(now.strftime("%I : %M"))
        self.popover_content.date_label.set_label(now.strftime("%A, %d %B"))
        GLib.timeout_add(1000, self.update_date_time)

File: src/modules/calendar.rs
use chrono::{DateTime, Local};
use relm4::gtk::prelude::*;
use relm4::{gtk, Component, ComponentParts, ComponentSender};
use std::{thread, time};

#[derive(Debug, Clone)]
pub struct Calendar {
    current_time: DateTime<Local>,
}

#[derive(Debug)]
pub enum CalendarInput {
    UpdateCurrentTime,
}

#[derive(Debug)]
pub enum CommandMsg {
    LocalTime(DateTime<Local>),
}

#[relm4::component(pub)]
impl Component for Calendar {
    type CommandOutput = CommandMsg;
    type Input = CalendarInput;
    type Output = ();
    type Init = ();

    view! {
        gtk::Box {
            gtk::Button {
                set_css_classes: &["calendar-toggle-button"],
                gtk::Label {
                    #[watch]
                    set_label: &model.current_time.format("%d %b %I:%M %p").to_string()
                },
                connect_clicked[popover] => move |_| {
                    popover.popup();
                }
            },
            #[name = "popover"]
            gtk::Popover {
                set_has_arrow: false,
                set_offset: (0, 8),
                gtk::Box {
                    set_orientation: gtk::Orientation::Vertical,
                    gtk::Label {
                        set_css_classes: &["title-1", "big-date-time"],
                        #[watch]
                        set_label: &model.current_time.format("%I : %M").to_string()
                    },
                    gtk::Label {
                        set_css_classes: &["title-4", "full-date"],
                        #[watch]
                        set_label: &model.current_time.format("%A, %d %B").to_string()
                    },
                    gtk::Calendar {}
                }
            }
        }
    }

    /// Initialize the UI and model.
    fn init(
        _config: Self::Init,
        root_box: &Self::Root,
        sender: ComponentSender<Self>,
    ) -> relm4::ComponentParts<Self> {
        let model = Calendar {
            current_time: Local::now(),
        };

        let widgets = view_output!();

        let _ = sender.input_sender().send(CalendarInput::UpdateCurrentTime);

        ComponentParts { model, widgets }
    }

    /// Update model based on message
    fn update(&mut self, msg: Self::Input, sender: ComponentSender<Self>, _: &Self::Root) {
        match msg {
            CalendarInput::UpdateCurrentTime => {
                sender.oneshot_command(async {
                    thread::sleep(time::Duration::from_secs(1));
                    // Run async background task
                    CommandMsg::LocalTime(Local::now())
                });
            }
        }
    }

    // Update model based of command output
    fn update_cmd(
        &mut self,
        message: Self::CommandOutput,
        sender: ComponentSender<Self>,
        _: &Self::Root,
    ) {
        match message {
            CommandMsg::LocalTime(time) => {
                self.current_time = time;
                let _ = sender.input_sender().send(CalendarInput::UpdateCurrentTime);
            }
        }
    }
}
